src/basset.cpp

```diff
@@ -190,79 +160,107 @@ void Regex::report(int errcode) const {
   CompilationDatabase &operator=(const CompilationDatabase &) = delete;
   CompilationDatabase &operator=(CompilationDatabase &&) = delete;
-  ~CompilationDatabase() override;
-  using ofstream::operator!;
+  ~CompilationDatabase() = default;
+
+  bool load();
+  bool save();
+
   void add(const string &directory, const vector<string> &command);
 
 private:
-  [[nodiscard]] static bool is_source_file(const string &argument);
-  bool first{true};
+  const string filename;
+  const IsSourceFileFunc is_source_file;
+
+  static string make_index_key(const string &directory,
+                               const string &filename) {
+    return canonical(path(directory) / filename);
+  }
+
+  // maps canonized source file path to (directory, arguments) pair
+  unordered_map<string, pair<string, vector<string>>> index;
 };
 
-template <typename... Args>
-CompilationDatabase::CompilationDatabase(Args &&...args)
-    : ofstream(std::forward<Args>(args)...) {
-  *this << '[';
+CompilationDatabase::CompilationDatabase(string filename,
+                                         IsSourceFileFunc is_source_file)
+    : filename(std::move(filename)), is_source_file(std::move(is_source_file)) {
 }
 
-void CompilationDatabase::add(const string &directory,
-                              const vector<string> &command) {
-  vector<string> files;
+bool CompilationDatabase::load() {
+  ifstream ifs(filename);
+  if (!ifs) {
+    return true;
+  }
 
-  for (auto &&argument : command) {
-    if (is_source_file(argument)) {
-      files.emplace_back(argument);
+  try {
+    auto json = nlohmann::json::parse(ifs);
+
+    // load existing database with removing entries for no longer existing
+    // sources
+    // (https://github.com/rizsotto/Bear/wiki/Features#append-to-existing-jspn-cdb)
+    for (auto it = json.begin(); it != json.end(); ++it) {
+      const auto directory = (*it)["directory"].get<string>();
+      const auto file = (*it)["file"].get<string>();
+
+      if (exists(path(directory) / file)) {
+        index[make_index_key(directory, file)] =
+            make_pair(directory, (*it)["arguments"]);
+      }
     }
+  } catch (const nlohmann::json::parse_error &e) {
+    cerr << "json error: " << e.what() << '\n';
+    return false;
   }
 
-  for (const auto &file : files) {
-    if (first) {
-      first = false;
-    } else {
-      *this << ',';
-    }
+  return true;
+}
 
-    *this << "\n"
-             "  {\n"
-             // clang-format off
-             "    \"directory\": \"" << json_escape(directory) << "\",\n"
-             // clang-format on
-             "    \"arguments\": [";
+bool CompilationDatabase::save() {
+  ofstream ofs(filename);
+  if (!ofs) {
+    return false;
+  }
 
-    bool first_arg{true};
+  auto json = nlohmann::json::array();
+  for (const auto &kv : index) {
+    json += nlohmann::json::object({{"directory", kv.second.first},
+                                    {"file", kv.first},
```

Comment:
As far as I understand, using key as `"file"` will have an effect that it will always be an absolute path, right?